README
NAME
    HTTP::Tiny - A small, simple, correct HTTP/1.1 client

VERSION
    version 0.053

SYNOPSIS
        use HTTP::Tiny;

        my $response = HTTP::Tiny->new->get('http://example.com/');

        die "Failed!\n" unless $response->{success};

        print "$response->{status} $response->{reason}\n";

        while (my ($k, $v) = each %{$response->{headers}}) {
            for (ref $v eq 'ARRAY' ? @$v : $v) {
                print "$k: $_\n";
            }
        }

        print $response->{content} if length $response->{content};

DESCRIPTION
    This is a very simple HTTP/1.1 client, designed for doing simple
    requests without the overhead of a large framework like LWP::UserAgent.

    It is more correct and more complete than HTTP::Lite. It supports
    proxies and redirection. It also correctly resumes after EINTR.

    If IO::Socket::IP 0.25 or later is installed, HTTP::Tiny will use it
    instead of IO::Socket::INET for transparent support for both IPv4 and
    IPv6.

    Cookie support requires HTTP::CookieJar or an equivalent class.

METHODS
  new
        $http = HTTP::Tiny->new( %attributes );

    This constructor returns a new HTTP::Tiny object. Valid attributes
    include:

    *   "agent" — A user-agent string (defaults to 'HTTP-Tiny/$VERSION'). If
        "agent" — ends in a space character, the default user-agent string
        is appended.

    *   "cookie_jar" — An instance of HTTP::CookieJar — or equivalent class
        that supports the "add" and "cookie_header" methods

    *   "default_headers" — A hashref of default headers to apply to
        requests

    *   "local_address" — The local IP address to bind to

    *   "keep_alive" — Whether to reuse the last connection (if for the same
        scheme, host and port) (defaults to 1)

    *   "max_redirect" — Maximum number of redirects allowed (defaults to 5)

    *   "max_size" — Maximum response size (only when not using a data
        callback). If defined, responses larger than this will return an
        exception.

    *   "http_proxy" — URL of a proxy server to use for HTTP connections
        (default is $ENV{http_proxy} — if set)

    *   "https_proxy" — URL of a proxy server to use for HTTPS connections
        (default is $ENV{https_proxy} — if set)

    *   "proxy" — URL of a generic proxy server for both HTTP and HTTPS
        connections (default is $ENV{all_proxy} — if set)

    *   "no_proxy" — List of domain suffixes that should not be proxied.
        Must be a comma-separated string or an array reference. (default is
        $ENV{no_proxy} —)

    *   "timeout" — Request timeout in seconds (default is 60)

    *   "verify_SSL" — A boolean that indicates whether to validate the SSL
        certificate of an "https" — connection (default is false)

    *   "SSL_options" — A hashref of "SSL_*" — options to pass through to
        IO::Socket::SSL

    Passing an explicit "undef" for "proxy", "http_proxy" or "https_proxy"
    will prevent getting the corresponding proxies from the environment.

    Exceptions from "max_size", "timeout" or other errors will result in a
    pseudo-HTTP status code of 599 and a reason of "Internal Exception". The
    content field in the response will contain the text of the exception.

    The "keep_alive" parameter enables a persistent connection, but only to
    a single destination scheme, host and port. Also, if any
    connection-relevant attributes are modified, or if the process ID or
    thread ID change, the persistent connection will be dropped. If you want
    persistent connections across multiple destinations, use multiple
    HTTP::Tiny objects.

    See "SSL SUPPORT" for more on the "verify_SSL" and "SSL_options"
    attributes.

  get|head|put|post|delete
        $response = $http->get($url);
        $response = $http->get($url, \%options);
        $response = $http->head($url);

    These methods are shorthand for calling "request()" for the given
    method. The URL must have unsafe characters escaped and international
    domain names encoded. See "request()" for valid options and a
    description of the response.

    The "success" field of the response will be true if the status code is
    2XX.

  post_form
        $response = $http->post_form($url, $form_data);
        $response = $http->post_form($url, $form_data, \%options);

    This method executes a "POST" request and sends the key/value pairs from
    a form data hash or array reference to the given URL with a
    "content-type" of "application/x-www-form-urlencoded". If data is
    provided as an array reference, the order is preserved; if provided as a
    hash reference, the terms are sorted on key and value for consistency.
    See documentation for the "www_form_urlencode" method for details on the
    encoding.

    The URL must have unsafe characters escaped and international domain
    names encoded. See "request()" for valid options and a description of
    the response. Any "content-type" header or content in the options
    hashref will be ignored.

    The "success" field of the response will be true if the status code is
    2XX.

  mirror
        $response = $http->mirror($url, $file, \%options)
        if ( $response->{success} ) {
            print "$file is up to date\n";
        }

    Executes a "GET" request for the URL and saves the response body to the
    file name provided. The URL must have unsafe characters escaped and
    international domain names encoded. If the file already exists, the
    request will include an "If-Modified-Since" header with the modification
    timestamp of the file. You may specify a different "If-Modified-Since"
    header yourself in the "$options->{headers}" hash.

    The "success" field of the response will be true if the status code is
    2XX or if the status code is 304 (unmodified).

    If the file was modified and the server response includes a properly
    formatted "Last-Modified" header, the file modification time will be
    updated accordingly.

  request
        $response = $http->request($method, $url);
        $response = $http->request($method, $url, \%options);

    Executes an HTTP request of the given method type ('GET', 'HEAD',
    'POST', 'PUT', etc.) on the given URL. The URL must have unsafe
    characters escaped and international domain names encoded.

    If the URL includes a "user:password" stanza, they will be used for
    Basic-style authorization headers. (Authorization headers will not be
    included in a redirected request.) For example:

        $http->request('GET', 'http://Aladdin:open sesame@example.com/');

    If the "user:password" stanza contains reserved characters, they must be
    percent-escaped:

        $http->request('GET', 'http://john%40example.com:password@example.com/');

    A hashref of options may be appended to modify the request.

    Valid options are:

    *   "headers" — A hashref containing headers to include with the
        request. If the value for a header is an array reference, the header
        will be output multiple times with each value in the array. These
        headers over-write any default headers.

    *   "content" — A scalar to include as the body of the request OR a code
        reference that will be called iteratively to produce the body of the
        request

    *   "trailer_callback" — A code reference that will be called if it
        exists to provide a hashref of trailing headers (only used with
        chunked transfer-encoding)

    *   "data_callback" — A code reference that will be called for each
        chunks of the response body received.

    The "Host" header is generated from the URL in accordance with RFC 2616.
    It is a fatal error to specify "Host" in the "headers" option. Other
    headers may be ignored or overwritten if necessary for transport
    compliance.

    If the "content" option is a code reference, it will be called
    iteratively to provide the content body of the request. It should return
    the empty string or undef when the iterator is exhausted.

    If the "content" option is the empty string, no "content-type" or
    "content-length" headers will be generated.

    If the "data_callback" option is provided, it will be called iteratively
    until the entire response body is received. The first argument will be a
    string containing a chunk of the response body, the second argument will
    be the in-progress response hash reference, as described below. (This
    allows customizing the action of the callback based on the "status" or
    "headers" received prior to the content body.)

    The "request" method returns a hashref containing the response. The
    hashref will have the following keys:

    *   "success" — Boolean indicating whether the operation returned a 2XX
        status code

    *   "url" — URL that provided the response. This is the URL of the
        request unless there were redirections, in which case it is the last
        URL queried in a redirection chain

    *   "status" — The HTTP status code of the response

    *   "reason" — The response phrase returned by the server

    *   "content" — The body of the response. If the response does not have
        any content or if a data callback is provided to consume the
        response body, this will be the empty string

    *   "headers" — A hashref of header fields. All header field names will
        be normalized to be lower case. If a header is repeated, the value
        will be an arrayref; it will otherwise be a scalar string containing
        the value

    On an exception during the execution of the request, the "status" field
    will contain 599, and the "content" field will contain the text of the
    exception.

  www_form_urlencode
        $params = $http->www_form_urlencode( $data );
        $response = $http->get("http://example.com/query?$params");

    This method converts the key/value pairs from a data hash or array
    reference into a "x-www-form-urlencoded" string. The keys and values
    from the data reference will be UTF-8 encoded and escaped per RFC 3986.
    If a value is an array reference, the key will be repeated with each of
    the values of the array reference. If data is provided as a hash
    reference, the key/value pairs in the resulting string will be sorted by
    key and value for consistent ordering.

SSL SUPPORT
    Direct "https" connections are supported only if IO::Socket::SSL 1.56 or
    greater and Net::SSLeay 1.49 or greater are installed. An exception will
    be thrown if new enough versions of these modules are not installed or
    if the SSL encryption fails. An "https" connection may be made via an
    "http" proxy that supports the CONNECT command (i.e. RFC 2817). You may
    not proxy "https" via a proxy that itself requires "https" to
    communicate.

    SSL provides two distinct capabilities:

    *   Encrypted communication channel

    *   Verification of server identity

    By default, HTTP::Tiny does not verify server identity.

    Server identity verification is controversial and potentially tricky
    because it depends on a (usually paid) third-party Certificate Authority
    (CA) trust model to validate a certificate as legitimate. This
    discriminates against servers with self-signed certificates or
    certificates signed by free, community-driven CA's such as CAcert.org
    <http://cacert.org>.

    By default, HTTP::Tiny does not make any assumptions about your trust
    model, threat level or risk tolerance. It just aims to give you an
    encrypted channel when you need one.

    Setting the "verify_SSL" attribute to a true value will make HTTP::Tiny
    verify that an SSL connection has a valid SSL certificate corresponding
    to the host name of the connection and that the SSL certificate has been
    verified by a CA. Assuming you trust the CA, this will protect against a
    man-in-the-middle attack
    <http://en.wikipedia.org/wiki/Man-in-the-middle_attack>. If you are
    concerned about security, you should enable this option.

    Certificate verification requires a file containing trusted CA
    certificates. If the Mozilla::CA module is installed, HTTP::Tiny will
    use the CA file included with it as a source of trusted CA's. (This
    means you trust Mozilla, the author of Mozilla::CA, the CPAN mirror
    where you got Mozilla::CA, the toolchain used to install it, and your
    operating system security, right?)

    If that module is not available, then HTTP::Tiny will search several
    system-specific default locations for a CA certificate file:

    *   /etc/ssl/certs/ca-certificates.crt

    *   /etc/pki/tls/certs/ca-bundle.crt

    *   /etc/ssl/ca-bundle.pem

    An exception will be raised if "verify_SSL" is true and no CA
    certificate file is available.

    If you desire complete control over SSL connections, the "SSL_options"
    attribute lets you provide a hash reference that will be passed through
    to "IO::Socket::SSL::start_SSL()", overriding any options set by
    HTTP::Tiny. For example, to provide your own trusted CA file:

        SSL_options => {
            SSL_ca_file => $file_path,
        }

    The "SSL_options" attribute could also be used for such things as
    providing a client certificate for authentication to a server or
    controlling the choice of cipher used for the SSL connection. See
    IO::Socket::SSL documentation for details.

PROXY SUPPORT
    HTTP::Tiny can proxy both "http" and "https" requests. Only Basic proxy
    authorization is supported and it must be provided as part of the proxy
    URL: "http://user:pass@proxy.example.com/".

    HTTP::Tiny supports the following proxy environment variables:

    *   http_proxy or HTTP_PROXY

    *   https_proxy or HTTPS_PROXY

    *   all_proxy or ALL_PROXY

    If the "REQUEST_METHOD" environment variable is set, then this might be
    a CGI process and "HTTP_PROXY" would be set from the "Proxy:" header,
    which is a security risk. If "REQUEST_METHOD" is set, "HTTP_PROXY" (the
    upper case variant only) is ignored.

    Tunnelling "https" over an "http" proxy using the CONNECT method is
    supported. If your proxy uses "https" itself, you can not tunnel "https"
    over it.

    Be warned that proxying an "https" connection opens you to the risk of a
    man-in-the-middle attack by the proxy server.

    The "no_proxy" environment variable is supported in the format of a
    comma-separated list of domain extensions proxy should not be used for.

    Proxy arguments passed to "new" will override their corresponding
    environment variables.

LIMITATIONS
    HTTP::Tiny is *conditionally compliant* with the HTTP/1.1 specifications
    <http://www.w3.org/Protocols/>:

    *   "Message Syntax and Routing" [RFC7230]

    *   "Semantics and Content" [RFC7231]

    *   "Conditional Requests" [RFC7232]

    *   "Range Requests" [RFC7233]

    *   "Caching" [RFC7234]

    *   "Authentication" [RFC7235]

    It attempts to meet all "MUST" requirements of the specification, but
    does not implement all "SHOULD" requirements. (Note: it was developed
    against the earlier RFC 2616 specification and may not yet meet the
    revised RFC 7230-7235 spec.)

    Some particular limitations of note include:

    *   HTTP::Tiny focuses on correct transport. Users are responsible for
        ensuring that user-defined headers and content are compliant with
        the HTTP/1.1 specification.

    *   Users must ensure that URLs are properly escaped for unsafe
        characters and that international domain names are properly encoded
        to ASCII. See URI::Escape, URI::_punycode and Net::IDN::Encode.

    *   Redirection is very strict against the specification. Redirection is
        only automatic for response codes 301, 302 and 307 if the request
        method is 'GET' or 'HEAD'. Response code 303 is always converted
        into a 'GET' redirection, as mandated by the specification. There is
        no automatic support for status 305 ("Use proxy") redirections.

    *   There is no provision for delaying a request body using an "Expect"
        header. Unexpected "1XX" responses are silently ignored as per the
        specification.

    *   Only 'chunked' "Transfer-Encoding" is supported.

    *   There is no support for a Request-URI of '*' for the 'OPTIONS'
        request.

    Despite the limitations listed above, HTTP::Tiny is considered
    feature-complete. New feature requests should be directed to
    HTTP::Tiny::UA.

SEE ALSO
    *   HTTP::Tiny::UA - Higher level UA features for HTTP::Tiny

    *   HTTP::Thin - HTTP::Tiny wrapper with HTTP::Request/HTTP::Response
        compatibility

    *   HTTP::Tiny::Mech - Wrap WWW::Mechanize instance in HTTP::Tiny
        compatible interface

    *   IO::Socket::IP - Required for IPv6 support

    *   IO::Socket::SSL - Required for SSL support

    *   LWP::UserAgent - If HTTP::Tiny isn't enough for you, this is the
        "standard" way to do things

    *   Mozilla::CA - Required if you want to validate SSL certificates

    *   Net::SSLeay - Required for SSL support

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <https://github.com/chansen/p5-http-tiny/issues>. You will be notified
    automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/chansen/p5-http-tiny>

      git clone https://github.com/chansen/p5-http-tiny.git

AUTHORS
    *   Christian Hansen <chansen@cpan.org>

    *   David Golden <dagolden@cpan.org>

CONTRIBUTORS
    *   Alan Gardner <gardner@pythian.com>

    *   Alessandro Ghedini <al3xbio@gmail.com>

    *   Brad Gilbert <bgills@cpan.org>

    *   Chris Nehren <apeiron@cpan.org>

    *   Chris Weyl <cweyl@alumni.drew.edu>

    *   Claes Jakobsson <claes@surfar.nu>

    *   Clinton Gormley <clint@traveljury.com>

    *   Craig Berry <cberry@cpan.org>

    *   David Mitchell <davem@iabyn.com>

    *   Dean Pearce <pearce@pythian.com>

    *   Edward Zborowski <ed@rubensteintech.com>

    *   James Raspass <jraspass@gmail.com>

    *   Jess Robinson <castaway@desert-island.me.uk>

    *   Lukas Eklund <leklund@gmail.com>

    *   Martin J. Evans <mjegh@ntlworld.com>

    *   Martin-Louis Bright <mlbright@gmail.com>

    *   Mike Doherty <doherty@cpan.org>

    *   Petr Písař <ppisar@redhat.com>

    *   Serguei Trouchelle <stro@cpan.org>

    *   Sören Kornetzki <soeren.kornetzki@delti.com>

    *   Syohei YOSHIDA <syohex@gmail.com>

    *   Tom Hukins <tom@eborcom.com>

    *   Tony Cook <tony@develop-help.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Christian Hansen.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release notes for HTTP-Tiny

0.053     2014-12-11 23:42:17-05:00 America/New_York

    [FIXED]

    - Defended tests against HTTP_PROXY set in the environment

0.052     2014-12-11 15:23:54-05:00 America/New_York

    [CHANGED]

    - Proxy allowed from environment variable HTTP_PROXY (uppercase) unless
      REQUEST_METHOD is also set.

0.051     2014-11-17 22:58:44-05:00 America/New_York

    [FIXED]

    - Checks for threads without loading threads.pm

0.050     2014-09-23 15:30:18-04:00 America/New_York

    [FIXED]

    - Fixed CONNECT requests for some proxies

0.049     2014-09-02 11:20:07-04:00 America/New_York

    [FIXED]

    - 'keep_alive' is now fork-safe and thread-safe

0.048     2014-08-21 13:19:51-04:00 America/New_York

    [FIXED]

    - Protected proxy tests from ALL_PROXY in the environment

0.047     2014-07-29 14:09:05-04:00 America/New_York

    [CHANGED]

    - Updated Mozilla::CA module recommendation version to 20130114

    [FIXED]

    - Fixed t/00-report-prereqs.t when CPAN::Meta is not installed

0.046     2014-07-21 10:32:32-04:00 America/New_York

    [FIXED]

    - Empty header fields are now allowed; headers with the 'undef' value
      will be rendered as an empty header.

    [DOCUMENTED]

    - Updated HTTP/1.1 spec description from RFC 2616 to RFC 7230-7235

0.045     2014-07-19 23:17:28-04:00 America/New_York (TRIAL RELEASE)

    [FIXED]

    - Fixed t/002_croakage.t for various operating systems.

0.044     2014-07-16 23:46:09-04:00 America/New_York

    [CHANGED]

    - Providing a custom 'Host' header is now a fatal exception.  Previously, it
      was silently ignored, as the RFC mandates that Host be set from the
      URL, but ignoring it could lead to unexpected, confusing errors.

    - optimized URL splitting

    - Passing 'undef' for any proxy attribute will prevent HTTP::Tiny from
      setting the proxy from the environment.

0.043     2014-02-20 20:40:23-05:00 America/New_York

    [FIXED]

    - Does not send absolute request URI when tunneling SSL via proxy

    - Fixes regression in setting host name to verify SSL

    - Protects tests from https_proxy and all_proxy when doing mock testing

0.042     2014-02-18 11:23:17EST-0500 America/New_York

    [ADDED]

    - If IO::Socket::IP 0.25+ is installed, HTTP::Tiny will use it for
      transparent IPv4 or IPv6 support.

0.041     2014-02-17 13:07:54-05:00 America/New_York

    [no code change, only an amended Changes file]

    [INCOMPATIBLE CHANGES (from 0.039)]

    - The 'proxy' attribute no longer takes precedence over the
      'http_proxy' environment variable.  With the addition of http_proxy
      and https_proxy attributes (and corresponding environment variable
      defaults), the legacy 'proxy' attribute now maps to the
      all_proxy/ALL_PROXY environment variable and only takes effect when
      other proxy attributes are not defined.

    [ADDED (since 0.039)]

    - Added 'keep_alive' attribute for single-server persistent connections
      (Clinton Gormley)

    - Added support for Basic authorization with proxies

    - Added support for https proxies via CONNECT

    [FIXED (since 0.039)]

    - Requests are made with one less write for lower latency (Martin
      Evans)

0.040     2014-02-17 13:02:47-05:00 America/New_York

    [INCOMPATIBLE CHANGES]

    - The 'proxy' attribute no longer takes precedence over the
      'http_proxy' environment variable.  With the addition of http_proxy
      and https_proxy attributes (and corresponding environment variable
      defaults), the legacy 'proxy' attribute now maps to the
      all_proxy/ALL_PROXY environment variable and only takes effect when
      other proxy attributes are not defined.

    [ADDED]

    - Added support for Basic authorization with proxies

    - Added support for https proxies via CONNECT

0.039     2013-11-27 19:48:29 America/New_York

    [FIXED]

    - Temporary file creating during mirror() is now opened with O_EXCL
      for added security

0.038     2013-11-18 12:56:26 America/New_York

    [FIXED]

    - Fixed a bug where authentication parameters in the URL would override
      an existing Authorization header

0.037     2013-10-28 13:26:21 America/New_York

    [FIXED]

    - Basic authentication in the URL is now unescaped before being encoded
      into the authentication header

    [DOCUMENTED]

    - Added HTTP::Tiny::UA to SEE ALSO and suggested it as the appropriate
      place for new features

0.036     2013-09-25 12:10:06 America/New_York

    [FIXED]

    - Compile test could hang on Windows

    [PREREQS]

    - Dropped configure_requires for ExtUtils::MakeMaker to 6.17

    [META]

    - Updated support files

0.035     2013-09-10 12:29:28 America/New_York

  [CHANGED]

  - Encoded from data from 'post_form' preserves term order if data is
    provided as an array reference.  (They are still sorted for consistency
    if provided as a hash reference.)

0.034     2013-06-26 19:02:25 America/New_York

  [ADDED]

  - Added support for 'Basic' authorization from
    user:password parameters in the URL

0.033     2013-06-21 06:26:51 America/New_York

  [FIXED]

  - Modifying the 'agent' attribute with the accessor will append the
    default agent string, just like setting it during construction

0.032     2013-06-20 11:41:24 America/New_York

  [ADDED]

  - Added 'no_proxy' attribute, defaulting to $ENV{no_proxy}

0.031     2013-06-16 23:18:18 America/New_York

  [FIXED]

  - Fixed bug receiving 0-length content bodies

0.030     2013-06-13 11:46:15 America/New_York

  [FIXED]

  - Requests with the empty string as body content no longer generate
    'content-type' and 'content-length' headers.

0.029     2013-04-17 13:49:07 America/New_York

  [FIXED]

  - Checks for new enough OpenSSL library before using SNI (otherwise
    IO::Socket::SSL throws warnings)

0.028     2013-03-05 14:11:57 America/New_York

  [SUPPORT]

  - Fix repository/issue links to reflect proper repo name

t/Util.pm
#
# This file is part of HTTP-Tiny
#
# This software is copyright (c) 2011 by Christian Hansen.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package t::Util;

use strict;
use warnings;

use IO::File q[SEEK_SET];
use IO::Dir;

BEGIN {
    our @EXPORT_OK = qw(
        rewind
        tmpfile
        dir_list
        slurp
        parse_case
        hashify
        sort_headers
        connect_args
        clear_socket_source
        set_socket_source
        monkey_patch
        $CRLF
        $LF
    );

    require Exporter;
    *import = \&Exporter::import;
}

our $CRLF = "\x0D\x0A";
our $LF   = "\x0A";

sub rewind(*) {
    seek($_[0], 0, SEEK_SET)
      || die(qq/Couldn't rewind file handle: '$!'/);
}

sub tmpfile {
    my $fh = IO::File->new_tmpfile
      || die(qq/Couldn't create a new temporary file: '$!'/);

    binmode($fh)
      || die(qq/Couldn't binmode temporary file handle: '$!'/);

    if (@_) {
        print({$fh} @_)
          || die(qq/Couldn't write to temporary file handle: '$!'/);

        seek($fh, 0, SEEK_SET)
          || die(qq/Couldn't rewind temporary file handle: '$!'/);
    }

    return $fh;
}

sub dir_list {
    my ($dir, $filter) = @_;
    $filter ||= qr/./;
    my $d = IO::Dir->new($dir)
        or return;
    return map { "$dir/$_" } sort grep { /$filter/ } grep { /^[^.]/ } $d->read;
}

sub slurp (*) {
    my ($fh) = @_;

    rewind($fh);

    binmode($fh)
      || die(qq/Couldn't binmode file handle: '$!'/);

    my $exp = -s $fh;
    my $buf = do { local $/; <$fh> };
    my $got = length $buf;

    ($exp == $got)
      || die(qq[I/O read mismatch (expexted: $exp got: $got)]);

    return $buf;
}

sub parse_case {
    my ($case) = @_;
    my %args;
    my $key = '';
    for my $line ( split "\n", $case ) {
        chomp $line;
        if ( substr($line,0,1) eq q{ } ) {
            $line =~ s/^\s+//;
            push @{$args{$key}}, $line;
        }
        else {
            $key = $line;
        }
    }
    return \%args;
}

sub hashify {
    my ($lines) = @_;
    return unless $lines;
    my %hash;
    for my $line ( @$lines ) {
        my ($k,$v) = ($line =~ m{^([^:]+): (.*)$}g);
        $hash{$k} = $v;
    }
    return %hash;
}

sub sort_headers {
    my ($text) = shift;
    my @lines = split /$CRLF/, $text;
    my $request = shift(@lines) || '';
    my @headers;
    while (my $line = shift @lines) {
        last unless length $line;
        push @headers, $line;
    }
    @headers = sort @headers;
    return join($CRLF, $request, @headers, '', @lines);
}

{
    my (@req_fh, @res_fh, $monkey_host, $monkey_port);

    sub clear_socket_source {
        @req_fh = ();
        @res_fh = ();
    }

    sub set_socket_source {
        my ($req_fh, $res_fh) = @_;
        push @req_fh, $req_fh;
        push @res_fh, $res_fh;
    }

    sub connect_args { return ($monkey_host, $monkey_port) }

    sub monkey_patch {
        no warnings qw/redefine once/;
        *HTTP::Tiny::Handle::can_read = sub {1};
        *HTTP::Tiny::Handle::can_write = sub {1};
        *HTTP::Tiny::Handle::connect = sub {
            my ($self, $scheme, $host, $port) = @_;
            $self->{host} = $monkey_host = $host;
            $self->{port} = $monkey_port = $port;
            $self->{fh} = shift @req_fh;
            return $self;
        };
        my $original_write_request = \&HTTP::Tiny::Handle::write_request;
        *HTTP::Tiny::Handle::write_request = sub {
            my ($self, $request) = @_;
            $original_write_request->($self, $request);
            $self->{fh} = shift @res_fh;
        };
        *HTTP::Tiny::Handle::close = sub { 1 }; # don't close our temps
    }
}

1;


# vim: et ts=4 sts=4 sw=4:


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.17;



my %WriteMakefileArgs = (
  "ABSTRACT" => "A small, simple, correct HTTP/1.1 client",
  "AUTHOR" => "Christian Hansen <chansen\@cpan.org>, David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "HTTP-Tiny",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "HTTP::Tiny",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Fcntl" => 0,
    "IO::Socket" => 0,
    "MIME::Base64" => 0,
    "Time::Local" => 0,
    "bytes" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Data::Dumper" => 0,
    "Exporter" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Basename" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "IO::Dir" => 0,
    "IO::File" => 0,
    "IO::Socket::INET" => 0,
    "IPC::Cmd" => 0,
    "Test::More" => "0.96",
    "open" => 0
  },
  "VERSION" => "0.050",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Data::Dumper" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "Fcntl" => 0,
  "File::Basename" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "IO::Dir" => 0,
  "IO::File" => 0,
  "IO::Socket" => 0,
  "IO::Socket::INET" => 0,
  "IPC::Cmd" => 0,
  "MIME::Base64" => 0,
  "Test::More" => "0.96",
  "Time::Local" => 0,
  "bytes" => 0,
  "open" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



